sqldb+invoices: fix incorrectly stored invoice expiries when using native SQL

bhandras opened the pull request:
Previously, when using the native schema, invoice expiries were incorrectly stored as 64-bit values (expiry in nanoseconds instead of seconds), causing overflow issues. Since we cannot determine the original values, we will set the expiries for existing invoices to 1 hour with this migration.

Based on https://github.com/lightningnetwork/lnd/pull/8854 only the last 5 commits are part of this PR.

saubyk commented:
> Since we cannot determine the original values

Question: Can you please elaborate on this? Is it because the precision will be lost when dividing the stored values by 1,000,000,000?

bhandras commented:
> > Since we cannot determine the original values
> 
> Question: Can you please elaborate on this? Is it because the precision will be lost when dividing the stored values by 1,000,000,000 to convert the nanoseconds into seconds?

Since 64-bit nanosecond values are truncated to 32 bits, we lose 32 bits of information, making it impossible to accurately determine the original intended values. Please note that this issue only affects newly stored invoices and only impacts users who are using native SQL.

bhandras commented on sqldb/sqlite.go:
```diff
@@ -26,6 +26,16 @@ const (
 	sqliteTxLockImmediate = "_txlock=immediate"
 )
 
+var (
+	// sqliteSchemaReplacements is a map of schema strings that need to be
+	// replaced for sqlite. This is needed because sqlite doesn't directly
+	// support the BIGINT type for primary keys, so we need to replace it
+	// with INTEGER.
+	sqliteSchemaReplacements = map[string]string{
+		"BIGINT PRIMARY KEY": "INTEGER PRIMARY KEY",
+	}
+)
```
Good question, I think we generally use a common SQL "dialect" for our schema definitions, but then we alter it slightly for both SQLite and Postgres. I believe we don't necessarily prefer one over the other, except for testing where a SQLite db is much faster to spin up. Perhaps we just reused code from Taproot Assets and these replaces come from there.

ProofOfKeags replied:
So is it fair to say then that the default queries are written against the SQL standard then? If that's the case should we change this commentary on the `postgresSchemaReplacements`

```
        // replaced for postgres. This is needed because we write the schemas
	// to work with sqlite primarily, and postgres has some differences.
```

bhandras replied:
Done, PTAL.

ellemouton reviewed (approved):
Nice - tACK - LGTM ✅ 

tested the following:

### Before this PR:
- Add invoice 
- see that expiry watcher uses correct expiry
- but incorrect expiry persisted
- on restart, invoice immediately expired & payer cannot pay

### Transitioning to this PR:

- on master, add the invoice. 
- restart LND with this PR
- see that persisted expiry is 3600
- invoice is loaded & payer can pay

My main comment is around potentially matching the default expiry that LND actually uses for new invoices though, which is 24 hrs, not 1 hr.

bhandras commented on sqldb/sqlite.go:
```diff
@@ -95,46 +105,43 @@ func NewSqliteStore(cfg *SqliteConfig, dbPath string) (*SqliteStore, error) {
+	}
+
+	// Now that the database is open, populate the database with our set of
+	// schemas based on our embedded in-memory file system.
```
Done (also for postgres).

bhandras commented on sqldb/migrations.go:
```diff
@@ -37,7 +92,15 @@ func applyMigrations(fs fs.FS, driver database.Driver, path,
 	}
-	err = sqlMigrate.Up()
+
+	migrationVersion, _, _ := sqlMigrate.Version()
```
Good catch ⚾ PTAL.

bhandras commented on sqldb/sqlc/migrations/000004_invoice_expiry_fix.up.sql:
```diff
@@ -0,0 +1,5 @@
+-- table. This is needed as previously we stored raw time.Duration values which
+-- are 64 bit integers and are used to express duration in nanoseconds however
+-- the intent is to store invoice expiry in seconds.
+UPDATE invoices SET expiry = 3600;
```
> one thing to note is that 3600 (1 hour ) is not the default used by LND. LND uses 24 hours (86400 seconds). See DefaultInvoiceExpiry (the one defined in invoicesrpc, not the one in zpay32)

Done, missed this, thank you!

> So cant it happen that perhaps any valid, unexpired, invoices currently in memory in the expiry watcher actually then dont get reloaded on startup since their expiry has been shortened significantly?

It can happen that we shorten intended expiries, but I'm not sure that we can do anything better. Thankfully the number of affected users and invoices should be pretty low since we don't have full migration of invoices yet, so only some new nodes are affected (who choose to use the native schema).

bhandras commented on sqldb/sqlite.go:
```diff
@@ -165,3 +165,29 @@ func NewTestSqliteDB(t *testing.T) *SqliteStore {
+	// an in mem version to speed up tests
+	dbFileName := filepath.Join(t.TempDir(), "tmp.db")
+	sqlDB, err := NewSqliteStore(&SqliteConfig{
+		SkipMigrations: true,
```
Given that this is for testing only imo it's better to be explicit since functional options also require developers to look up usage in the code and compiler won't help catching incorrect use. 

I agree though it's a nice way to hide complexity :) Happy to change it if you feel it's better!

bhandras reviewed (commented):
Thank you for the thorough review and insightful comments @ellemouton! Much appreciated :)

saubyk commented on docs/release-notes/release-notes-0.18.2.md:
```diff
@@ -86,6 +86,13 @@
+* [Fixed](https://github.com/lightningnetwork/lnd/pull/8854) pagination issues 
+  in SQL invoicedb queries.
+
+* [Migrate](https://github.com/lightningnetwork/lnd/pull/8855) incorrectly
```
It would be good to mention in the release notes as well that after the migration the default expiry of the invoices will be 24 Hours

bhandras replied:
Done, ptal.

saubyk reviewed (approved):
LGTM 💾

ellemouton commented:
@bhandras  - just picked this up while going through some route blinding stuff: I see that for AMP invoices, our default is actually 30 days (instead of 24 hours like as for non-amp invoices). Is that potentially something we want to take into account?

bhandras commented:
> @bhandras  - just picked this up while going through some route blinding stuff: I see that for AMP invoices, our default is actually 30 days (instead of 24 hours like as for non-amp invoices). Is that potentially something we want to take into account?

Good catch, thank you! Let me update the PR to also fix times for AMP invoices before we merge. Will re-req when done.

bhandras commented:
List of changes since last round:

- AMP invoices are migrated to 24h expiry
- Fixed some issues with the postgres test helpers
- Added coverage for AMP invoices as well as running the migration test for both sqlite and postgres

Diff: https://github.com/lightningnetwork/lnd/compare/a50d8bd58bbc08bd0f75761315a6551a502664bc..a779bc3f46b656f3be8b5bc7bc5e9d9b5b6a1938

ellemouton commented on sqldb/migrations_test.go:
```diff
@@ -0,0 +1,170 @@
+	_, err := db.InsertInvoice(ctxb, sqlc.InsertInvoiceParams{
+		Hash:               hash1,
+		PaymentAddr:        hash1,
+		PaymentRequestHash: hash1,
+		Expiry:             -123,
```
does this actually set it to NULL? surely it just sees this as `IsAmp: false`?

bhandras replied:
Great catch, fixed!

ellemouton reviewed (approved):
LGTM! 🚢
